c10/core/TensorImpl.cpp

```diff
@@ -81,20 +101,27 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
     // UndefinedTensorImpl is a singleton, so we skip logging it
     C10_LOG_API_USAGE_ONCE("tensor.create");
   }
-  // After we removed Autograd keys from globally enabled set, every Tensor must be created with
-  // a backend DispatchKey and an AutogradBackend key.
-  // We automatically add the corresponding autograd key to key_set_ so that backends can stay
-  // in the old way of only registering with backend key like DispatchKey::CPU.
-  if (c10::InferenceMode::is_enabled()) {
-    // See Note [Expected TLS state in InferenceMode] for why we don't add Autograd & InplaceOrView keys.
-    key_set_ = key_set;
+
+  bool inference_mode = c10::InferenceMode::is_enabled();
+
+  // Inference tensor doesn't have autograd related keys.
+  if (inference_mode) {
+    // See Note [Expected TLS state in InferenceMode] for why we exclude Autograd & InplaceOrView keys.
+    // Normally key_set only contains backend keys but we do the substraction
+    // here to make sure.
+    key_set_ = key_set - c10::autograd_dispatch_keyset_with_InplaceOrView;
```

Comment:
nit: as a point of hygiene, if a situation is supposed to be illegal, it's better to ASSERT it, rather than silently work around it, as it will help people identify if something has gone wrong in the system earlier rather than later